stock_reserve/model/stock_reserve.py

```diff
@@ -0,0 +1,181 @@
+        if picking_type_id and not res.get("location_id", False):
+            picking = self.env["stock.picking"].new(
+                {"picking_type_id": picking_type_id}
+            )
+            picking.onchange_picking_type()
+            res["location_id"] = picking.location_id.id
+        if "location_dest_id" in fields_list:
+            res["location_dest_id"] = self._default_location_dest_id()
+        if "product_uom_qty" in fields_list:
+            res["product_uom_qty"] = 1.0
+        return res
+
+    @api.model
+    def get_location_from_ref(self, ref):
+        """ Get a location from a xmlid if allowed
+        :param ref: tuple (module, xmlid)
+        """
+        try:
+            location = self.env.ref(ref, raise_if_not_found=True)
+            location.check_access_rule("read")
+            location_id = location.id
+        except (except_orm, ValueError):
+            location_id = False
+        return location_id
+
+    @api.model
+    def _default_picking_type_id(self):
+        ref = "stock.picking_type_out"
+        return self.env.ref(ref, raise_if_not_found=False).id
+
+    @api.model
+    def _default_location_dest_id(self):
+        ref = "stock_reserve.stock_location_reservation"
+        return self.get_location_from_ref(ref)
+
+    def reserve(self):
+        """ Confirm reservations
+
+        The reservation is done using the default UOM of the product.
+        A date until which the product is reserved can be specified.
+        """
+        self.write({"date_expected": fields.Datetime.now()})
+        self.mapped("move_id")._action_confirm(merge=False)
+        self.mapped("move_id.picking_id").action_assign()
+        return True
+
+    def release_reserve(self):
+        """
+        Release moves from reservation
+        """
+        moves = self.mapped("move_id")
+        moves._action_cancel()
+        # HACK: For avoiding to accumulate all reservations in the same picking
+        moves.write({"picking_id": False})
+        return True
+
+    @api.model
+    def release_validity_exceeded(self, ids=None):
+        """ Release all the reservation having an exceeded validity date """
+        domain = [
+            ("date_validity", "<", fields.date.today()),
+            ("state", "in", ["assigned", "confirmed"]),
+        ]
+        if ids:
+            domain.append(("id", "in", ids))
+        self.env["stock.reservation"].search(domain).release_reserve()
+        return True
+
+    def unlink(self):
+        """ Release the reservation before the unlink """
+        self.release_reserve()
+        return super().unlink()
+
+    @api.onchange("product_id")
+    def _onchange_product_id(self):
+        """ set product_uom and name from product onchange """
+        # save value before reading of self.move_id as this last one erase
+        # product_id value
+        self.move_id.write({"product_id": self.product_id.id})
+        self.move_id.onchange_product_id()
```

Comment:
I have some doubts about the way this is done:

1. Is it "safe" to do a `write` in an `onchange`? Than can end up triggering too many writes, or even cause some weird behavior. Could it maybe be converted to a `compute` and apply a similar behavior?
2. Is it still necessary to manually trigger the `onchange` in the `stock.move`? 

As I said, they are doubts, maybe this is the best solution, just wanted to confirm 🙂

Reply:
No, you can't do a write on an onchange. That's totally forbidden.

Reply:
On the second point, I think that's necessary, because the change is launched by code. If we just assign the `product_id` on the `stock.move`, we will not take the value of `name` and `product_uom`